File: game/atc-core/server/npc.lua
-- ATC NPC & Dynamic Population Runtime Bridge
-- Spawn/despawn NPCs, record behaviors, update crowd density.
-- All population state is server-authoritative.

local API_BASE  = ATC.Config.ApiBase  or 'http://localhost:3000'
local API_TOKEN = ATC.Config.ApiToken or ''
local SERVER_ID = GetCurrentServerId() or 'server-1'

local function apiPost(path, body, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'POST', json.encode(body), {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

local function apiGet(path, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'GET', '', {
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

-- ── Public SDK ────────────────────────────────────────────────────────────────

ATC.NPC = {}

--- Spawn a new NPC into the runtime.
--- @param zoneId      string
--- @param spawnNonce  string  idempotency key (use a UUID)
--- @param npcType     string  'civilian'|'pedestrian'|'ambient'|'scripted'|'emergency'
--- @param metadata    table|nil  arbitrary metadata
--- @param cb          function callback(status, npc)
function ATC.NPC.Spawn(zoneId, spawnNonce, npcType, metadata, cb)
  apiPost('/api/v1/npc/spawn', {
    zoneId        = zoneId,
    spawnNonce    = spawnNonce,
    npcType       = npcType or 'civilian',
    ownerServerId = SERVER_ID,
    metadata      = metadata or {},
  }, cb)
end

--- Despawn an NPC from the runtime.
--- @param npcId  string
--- @param reason string|nil  e.g. 'player_entered_zone', 'timeout'
--- @param cb     function callback(status)
function ATC.NPC.Despawn(npcId, reason, cb)
  apiPost('/api/v1/npc/despawn', {
    npcId         = npcId,
    reason        = reason or 'manual',
    ownerServerId = SERVER_ID,
  }, cb)
end

--- Send a heartbeat for an NPC to keep it alive.
--- @param npcId string
--- @param cb    function|nil callback(status)
function ATC.NPC.Heartbeat(npcId, cb)
  apiPost('/api/v1/npc/heartbeat', {
    npcId         = npcId,
    ownerServerId = SERVER_ID,
  }, cb)
end

--- Record an NPC behavior transition.
--- @param npcId    string
--- @param behavior string  e.g. 'idle', 'walking', 'fleeing', 'talking'
--- @param params   table|nil
--- @param cb       function|nil callback(status, record)
function ATC.NPC.RecordBehavior(npcId, behavior, params, cb)
  apiPost('/api/v1/npc/' .. npcId .. '/behavior', {
    npcId    = npcId,
    behavior = behavior,
    params   = params or {},
  }, cb)
end

--- Update crowd density for a zone.
--- @param zoneId         string
--- @param density        number  0.0–1.0
--- @param targetDensity  number|nil  0.0–1.0
--- @param activeNpcCount number|nil
--- @param cb             function|nil callback(status, crowd)
function ATC.NPC.UpdateCrowd(zoneId, density, targetDensity, activeNpcCount, cb)
  apiPost('/api/v1/npc/crowd', {
    zoneId         = zoneId,
    density        = density,
    targetDensity  = targetDensity or density,
    activeNpcCount = activeNpcCount or 0,
  }, cb)
end

--- Get crowd data for a zone.
--- @param zoneId string
--- @param cb     function callback(status, crowd)
function ATC.NPC.GetCrowd(zoneId, cb)
  apiGet('/api/v1/npc/crowd/' .. zoneId, cb)
end

--- Trigger stale NPC cleanup for this server.
--- @param cb function|nil callback(status, result)
function ATC.NPC.CleanupStale(cb)
  apiPost('/api/v1/npc/cleanup', {
    ownerServerId    = SERVER_ID,
    staleThresholdMs = 30000,
  }, cb)
end

-- ── Periodic heartbeat for all owned NPCs ─────────────────────────────────────

local ownedNpcs = {}

--- Register an NPC handle so the bridge sends automatic heartbeats.
--- @param npcId string
function ATC.NPC.TrackOwnership(npcId)
  ownedNpcs[npcId] = true
end

--- Unregister an NPC from automatic heartbeat tracking.
--- @param npcId string
function ATC.NPC.ReleaseOwnership(npcId)
  ownedNpcs[npcId] = nil
end

-- Heartbeat loop: every 10 seconds
CreateThread(function()
  while true do
    Wait(10000)
    for npcId, _ in pairs(ownedNpcs) do
      ATC.NPC.Heartbeat(npcId)
    end
  end
end)

-- Stale cleanup on resource start
AddEventHandler('onResourceStart', function(resourceName)
  if resourceName ~= GetCurrentResourceName() then return end
  Wait(5000)
  ATC.NPC.CleanupStale()
end)

File: apps/api/src/routes/npc.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import { requireCapability } from '../middleware/authorization.js'
import {
  spawnNpcSchema,
  despawnNpcSchema,
  recordNpcBehaviorSchema,
  npcHeartbeatSchema,
  updateCrowdDensitySchema,
  cleanupStaleNpcsSchema,
} from '@atc/operations'
import { NpcRuntimeError } from '@atc/npc-runtime'

function npcErrorToStatus(err: NpcRuntimeError): number {
  const name = err.constructor.name
  if (name === 'NpcAlreadyOwnedError' || name === 'NpcSpawnNonceConflictError') return 409
  if (name.endsWith('NotFoundError')) return 404
  return 500
}

const NOT_CONFIGURED = { error: 'NPC runtime not configured' }

export async function npcRoutes(
  fastify: FastifyInstance,
  opts: { ctx: AppContext },
): Promise<void> {
  const { ctx } = opts

  // ── Spawn NPC ─────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/spawn', {
    preHandler: requireCapability(ctx, 'npc:write'),
    handler: async (req, reply) => {
      if (!ctx.dynamicSpawnService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = spawnNpcSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const npcType = (parsed.data.npcType ?? 'civilian') as 'civilian' | 'pedestrian' | 'ambient' | 'scripted' | 'emergency'
        const result = await ctx.dynamicSpawnService.spawnNpc({
          spawnNonce:  parsed.data.spawnNonce,
          npcType,
          ...(parsed.data.zoneId !== undefined ? { zoneId: parsed.data.zoneId } : {}),
          ...(parsed.data.ownerServerId !== undefined ? { ownerServerId: parsed.data.ownerServerId } : {}),
          ...(parsed.data.metadata !== undefined ? { metadata: parsed.data.metadata } : {}),
        })
        return reply.status(201).send(result)
      } catch (err) {
        if (err instanceof NpcRuntimeError) return reply.status(npcErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Despawn NPC ───────────────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/despawn', {
    preHandler: requireCapability(ctx, 'npc:write'),
    handler: async (req, reply) => {
      if (!ctx.dynamicSpawnService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = despawnNpcSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        await ctx.dynamicSpawnService.despawnNpc(
          parsed.data.npcId,
          parsed.data.ownerServerId,
        )
        return reply.status(204).send()
      } catch (err) {
        if (err instanceof NpcRuntimeError) return reply.status(npcErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get NPC ───────────────────────────────────────────────────────────────────

  fastify.get('/api/v1/npc/:npcId', {
    preHandler: requireCapability(ctx, 'npc:read'),
    handler: async (req, reply) => {
      if (!ctx.npcRuntimeRepo) return reply.status(503).send(NOT_CONFIGURED)
      const { npcId } = req.params as { npcId: string }
      const npc = await ctx.npcRuntimeRepo.findById(npcId)
      if (!npc) return reply.status(404).send({ error: 'NpcNotFound' })
      return reply.send(npc)
    },
  })

  // ── NPC heartbeat ─────────────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/heartbeat', {
    preHandler: requireCapability(ctx, 'npc:write'),
    handler: async (req, reply) => {
      if (!ctx.npcRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = npcHeartbeatSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        await ctx.npcRuntimeService.heartbeat(parsed.data.npcId)
        return reply.status(204).send()
      } catch (err) {
        if (err instanceof NpcRuntimeError) return reply.status(npcErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Record NPC behavior ───────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/:npcId/behavior', {
    preHandler: requireCapability(ctx, 'npc:write'),
    handler: async (req, reply) => {
      if (!ctx.ambientBehaviorService) return reply.status(503).send(NOT_CONFIGURED)
      const { npcId } = req.params as { npcId: string }
      const parsed = recordNpcBehaviorSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.ambientBehaviorService.recordBehavior(
          npcId,
          parsed.data.behavior,
          parsed.data.params,
        )
        return reply.send(result)
      } catch (err) {
        if (err instanceof NpcRuntimeError) return reply.status(npcErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Update crowd density ──────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/crowd', {
    preHandler: requireCapability(ctx, 'npc:write'),
    handler: async (req, reply) => {
      if (!ctx.crowdSimulationService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = updateCrowdDensitySchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.crowdSimulationService.updateDensity(
          parsed.data.zoneId,
          parsed.data.density,
          parsed.data.targetDensity ?? parsed.data.density,
          parsed.data.activeNpcCount ?? 0,
        )
        return reply.send(result)
      } catch (err) {
        if (err instanceof NpcRuntimeError) return reply.status(npcErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get crowd density for zone ────────────────────────────────────────────────

  fastify.get('/api/v1/npc/crowd/:zoneId', {
    preHandler: requireCapability(ctx, 'npc:read'),
    handler: async (req, reply) => {
      if (!ctx.crowdSimulationService) return reply.status(503).send(NOT_CONFIGURED)
      const { zoneId } = req.params as { zoneId: string }
      const crowd = await ctx.crowdSimulationService.getCrowd(zoneId)
      if (!crowd) return reply.status(404).send({ error: 'CrowdNotFound' })
      return reply.send(crowd)
    },
  })

  // ── Cleanup stale NPCs ────────────────────────────────────────────────────────

  fastify.post('/api/v1/npc/cleanup', {
    preHandler: requireCapability(ctx, 'npc:admin'),
    handler: async (req, reply) => {
      if (!ctx.npcRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = cleanupStaleNpcsSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      const cleaned = await ctx.npcRuntimeService.reconcile([])
      return reply.send({ cleaned })
    },
  })
}
